data/gen.go

```diff
@@ -0,0 +1,301 @@
+package main
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+)
+
+const english_language_id = 9
+const national_dex_id = 1
+
+type data_pokemon struct {
+	Identifier     string
+	SpeciesId      int
+	Height         int
+	Weight         int
+	BaseExperience int
+
+	Name   string
+	NatDex uint16
+}
+
+type data_move_flags uint32
+
+type data_move struct {
+	Id             int
+	Identifier     string
+	SpeciesId      int
+	Height         int
+	Weight         int
+	BaseExperience int
+
+	Name     string
+	Power    int
+	PP       int
+	Accuracy int
+	Priority int
+	// see: move_targets.csv
+	Targets     int
+	DamageClass int
+	Effect      int
+	Flags       data_move_flags
+}
+
+func getCsvReader(path string) *csv.Reader {
+	log.Printf("Reading csv: %s\n", path)
+	file, err := os.Open(path)
+	if err != nil {
+		log.Panicln(err)
+	}
+	reader := csv.NewReader(file)
+	reader.ReuseRecord = true
+	reader.Read() // skip header line
+	return reader
+}
+
+func contains(s []int, e int) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func createCodeOutput() *os.File {
+	file, err := os.Create("pokedex_GEN.go")
+	if err != nil {
+		log.Panicln(err)
+	}
+	file.WriteString("// Code generated - DO NOT EDIT.\n")
+	file.WriteString("// Regenerate with `go generate`.\n\n")
+	file.WriteString("package pokemonbattlelib\n\n")
+	return file
+}
+
+func main() {
+	path, err := os.Getwd()
+	if err != nil {
+		log.Panicln(err)
+	}
+	log.Printf("current directory: %s\n", path)
+
+	// get all valid version ids
+	valid_version_groups := []int{}
+	version_groups_csv := getCsvReader("data/version_groups.csv")
+	for {
+		record, err := version_groups_csv.Read()
+		if err == io.EOF {
+			break
+		}
+		gen_id, err := strconv.Atoi(record[2])
+		if gen_id <= 4 {
```

Comment:
maybe we could have the `4` here be a constant, say `MOST_CURRENT_SUPPORTED_GEN`, at the top of the file so that if the library ever expands to support later generations the change will be easier to make